### PR TITLE
Update installer (resolves DEP-21)

### PR DESCRIPTION
**Pull request by jessebrennan**
This updates the installer with the three new options that are necessary to access the whitelist.

### CODE REVIEW

**Comment by jessebrennan** [comment]
I'm not really sure why there are merge conflicts, I might have done some git foo(bar) and screwed things up. @coverbeck could you review?

**Review by coverbeck — CHANGES_REQUESTED** [review]
1. Can you please resolve conflicts
2. You need to use these in prod.yml and dev.yml.

**Comment by coverbeck** [comment]
Your commit message should refer to the GitHub issue, not the JIRA, at least in the first line.

It would also be nice if the commit message gave a little more detail so somebody could know what changed without having to look at the code change.

File: install_bootstrap [hunk 1]
```diff
@@ -249,7 +249,15 @@ while [[ "${run_boardwalk^^}" != 'Y' &&  "${run_boardwalk^^}" != 'N' ]] ; do
 #            read -ep $'What is the email address for notification? \n' email_address
 #      fi
 
-      
+      project_name='project_name'
```
Comment (coverbeck) [inline install_bootstrap]: This will only come into play if there is an email_whitelist_name.

I think you should ask for the email_whitelist_name first. If the user does not enter one, then you shouldn't prompt for this nor for the email.

File: install_bootstrap [hunk 2]
```diff
@@ -249,7 +249,15 @@ while [[ "${run_boardwalk^^}" != 'Y' &&  "${run_boardwalk^^}" != 'N' ]] ; do
...
+      ask_question "What is the name of the project for this instance (e.g. Stage, Commons, AnVIL, etc.)? This name will be used for error messages." "$PROJECT_NAME" "project name" $project_name
+
+      contact_email='contact_email'
+      ask_question "What is a contact email in case there are problems authenticating?" "$CONTACT_EMAIL" "contact email" $contact_email
+
+      email_whitelist_name='email_whitelist_name'
+      ask_question "What is the name of the AWS Secret that contains the whitelist? (e.g. commons/dev/whitelist)" "$EMAIL_WHITELIST_NAME" "email whitelist name" $email_whitelist_name
```
Comment (coverbeck) [inline install_bootstrap]: We want to use this as a flag to indicate whether authentication is required to access the instance. The prompt should reflect that, e.g., something like like "leave blank if you do not want to enforce authorization".

File: install_bootstrap [hunk 3]
```diff
@@ -249,7 +249,15 @@ while [[ "${run_boardwalk^^}" != 'Y' &&  "${run_boardwalk^^}" != 'N' ]] ; do
 #            read -ep $'What is the email address for notification? \n' email_address
 #      fi
 
-      
+      project_name='project_name'
+      ask_question "What is the name of the project for this instance (e.g. Stage, Commons, AnVIL, etc.)? This name will be used for error messages." "$PROJECT_NAME" "project name" $project_name
+
+      contact_email='contact_email'
+      ask_question "What is a contact email in case there are problems authenticating?" "$CONTACT_EMAIL" "contact email" $contact_email
```
Comment (coverbeck) [inline install_bootstrap]: I would say `authorizing` instead of `authenticating`.

**Review by coverbeck — CHANGES_REQUESTED** [review]
There are two other things in the ticket:

1. Google client ids, but that is optional, so you can leave it out for now.
2. Generating a secret for cookie encryption/decryption. I think you should tackle that one. There is a generatePassword function in install_boostrap that you can invoke.

Also, can you please update the Collecting Information section of the README with a little more information about these.

I know you said you didn't like how the README.md presents information, so if you want to refactor it overall as part of this, that would be fine too... Or just fit it into the existing format, and create another ticket to update the README.

Thanks!

File: install_bootstrap (hunk 1 above)
Comment (jessebrennan) [reply install_bootstrap]: Good point. I've reorganized.

File: install_bootstrap (hunk 2 above)
Comment (jessebrennan) [reply install_bootstrap]: This is also fixed now.

File: install_bootstrap (hunk 3 above)
Comment (jessebrennan) [reply install_bootstrap]: Good catch!

File: boardwalk/conf/boardwalk.config.template [hunk 4]
```diff
@@ -1,6 +1,9 @@
 azul_s3_bucket={{ AZUL_S3_BUCKET }}
 user_aws_key_id={{ AWS_ACCESS_KEY_ID }}
 user_aws_secret_key={{ AWS_SECRET_ACCESS_KEY }}
+project_name={{ PROJECT_NAME }}
+contact_email={{ CONTACT_EMAIL }}
+email_whitelist_name={{ EMAIL_WHITELIST_NAME }}
```
Comment (coverbeck) [inline boardwalk/conf/boardwalk.config.template]: Don't you need to add `refresh_token_encrypt_key` in here as well?

File: README.md [hunk 5]
```diff
@@ -2,40 +2,94 @@
...
 
 ![Computational Genomics Platform architecture](docs/dcc-arch.png)
 
-These components are setup with the install process available in this repository:
+These components are setup with the install process available in this
+repository:
 
-* [Boardwalk](boardwalk/README.md): our file browsing portal on top of Redwood
+* [Boardwalk](boardwalk/README.md): our file browsing portal on top of
+  Redwood
```
Comment (coverbeck) [inline README.md]: I know this predates your changes (it was something that got overlooked), but can you instead say something like `...on top of the cgp-data-store`. Or maybe `on top of the Azul indexer`, as the UI doesn't directly interact with the data store.

File: README.md [hunk 6]
```diff
@@ -2,40 +2,94 @@
...
 
 * Ubuntu Server 16.04
 * r4.xlarge
 * 250GB disk
 
-For development work the following specifications have worked well in the past:
-* Ubuntu Server 16.04
-* m5.large
-* 60 GB disk
+For small scale development, a t2.medim with 20GB disks running Ubuntu
```
Comment (coverbeck) [inline README.md]: `medium`

File: README.md [hunk 7]
```diff
@@ -2,40 +2,94 @@
...
+   *Associate*, choose resource type "Instance", and choose your EC2
+   (e.g., searching by its name).
+7.  <a name="sshconnect"></a>In _EC2 Dashboard_ make your VM active by
+    clicking it. Then click _Connect_ on top. The example in that window
+    shows you how to ssh into your VM from a terminal.
+
+
+#### Make a routing to the VM
+
+You will need to make a URL to access the VM. On the AWS console, go to
```
Comment (coverbeck) [inline README.md]: This is nitpicky, but you can access the VM via SSH without a URL.... I would rephrase this a little bit, but I'm drawing a blank on how to best say it. :)

File: README.md [hunk 8]
```diff
@@ -44,94 +98,211 @@ Open inbound ports on your security group. Use the table below as a guide. Make
...
-* You will be asked to provide an host domain that points to your EC2 instance. You need to know the name of the domain but at the time of installation the domain (or _record set_) does not have to be configured in _Route 53_.
-* _Boardwalk_ has functionality to export metadata to Broad's FireCloud. In order to use it you need to provide your Google Cloud Platform credentials. Specifically you need the Google Client ID and the Google Client Secret (it's okay to leave the Google site verfication code empty). 
-* You need to input the [AWS IAM user](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_access-keys.html)'s _ access key ID_ and *secret_access_key*.
-* You need to have the name of the [S3 bucket you created earlier](#makebucket) handy as the install script will ask for it. The name you input has to be the exact name you gave it when you created the bucket.
-* All metadata reside in an Elasticsearch database. Make sure you have the domain name of that Elasticsearch instance handy.
-* Have the domain name of the _dos-dss server_ handy.
+Now you can begin installation. You will clone the repository on VM and
```
Comment (coverbeck) [inline README.md]: ...clone *this* repository...

File: README.md [hunk 9]
```diff
@@ -44,94 +98,211 @@ Open inbound ports on your security group. Use the table below as a guide. Make
...
+   missed it before, instructions to set this up are found on the
+   [Boardwalk deployment page](boardwalk/README.md#deployment).
+   _Boardwalk_ has functionality to export metadata to Broad's
+   FireCloud. In order to use it you need to provide your Google Cloud
+   Platform credentials.
+
+   You will also be asked for a Google site verfication code. For this
+   field you can just enter `NONE`.
+
+1. The bagit-firecloud-lambda is use for the **Export to Firecloud**
```
Comment (coverbeck) [inline README.md:226]: `used`

File: README.md [hunk 10]
```diff
@@ -44,94 +98,211 @@ Open inbound ports on your security group. Use the table below as a guide. Make
...
+   button.
+
+1. The [S3 bucket you created earlier](#makebucket) is what you should
+   use for this next step.
+
+1. The dos-dss server is needed next. Again you can use a prexisting one
```
Comment (coverbeck) [inline README.md]: I think you will always use an existing one.

File: README.md [hunk 11]
```diff
@@ -44,94 +98,211 @@ Open inbound ports on your security group. Use the table below as a guide. Make
...
+cause [Let's Encrypt](https://letsencrypt.org/) to issue fake SSL
+certificates, which won't exhaust your certificate's limit. Setting up
+*Boardwalk* to run in `dev` mode will first build then run the Docker
+containers `boardwalk_nginx`, `boardwalk_dcc-dashboard`,
+`boardwalk_dcc-dashboard-service`, and `boardwalk_boardwalk` from the
+images (see
+[here](https://github.com/DataBiosphere/cgp-deployment/blob/feature/update-readme/boardwalk/README.md#development-mode)
+for more details). In addition, the `nginx` image is built from the
+*nginx-dev* directory. If your work requires real SSL certificates
+during development, it is recommended to set up *Common* in `prod` mode,
```
Comment (coverbeck) [inline README.md:263]: Maybe put this at the very beginning of the section, i.e., even if running in dev mode, we recommend running Common in prod mode. Otherwise it's easy (for me) to overlook, and I think you ran into this too when you set up your first instance.

File: README.md [hunk 12]
```diff
@@ -44,94 +98,211 @@ Open inbound ports on your security group. Use the table below as a guide. Make
...
+1. The [S3 bucket you created earlier](#makebucket) is what you should
+   use for this next step.
+
+1. The dos-dss server is needed next. Again you can use a prexisting one
+   if this is just for development.
+
+### <a name="prod"></a> Installing in `prod` mode
 
-### Running the Installer
-Once you have collected the above information, clone the repository on VM and run the bootstrap script. Be sure to set your branch to `feature/commons` as these instruction are specific to this branch.
```
Comment (coverbeck) [inline README.md]: Optional: Maybe there should be a small section on why you would choose dev vs. prod mode. It's sort of in the details, but I don't think it's clear.  Or maybe the names dev and prod make it clear enough?

File: README.md [hunk 13]
```diff
@@ -44,94 +98,211 @@ Open inbound ports on your security group. Use the table below as a guide. Make
...
-* This [blog post](https://www.digitalocean.com/community/tutorials/how-to-remove-docker-images-containers-and-volumes) is helpful if you want to clean up previous images/containers/volumes.
+* This [blog
+  post](https://www.digitalocean.com/community/tutorials/how-to-remove-docker-images-containers-and-volumes)
+  is helpful if you want to clean up previous images/containers/volumes.
 
 ### To Do
 
 * the bootstrapper should install Java, Dockstore CLI
-* Consonance config.template includes hard-coded Consonance token, needs to be generated and written to .env file just like Beni does
+* Consonance config.template includes hard-coded Consonance token, needs
```
Comment (coverbeck) [inline README.md]: Please remove this, it was overlooked. This branch does not include Consonance.

File: install_bootstrap [hunk 14]
```diff
@@ -58,14 +58,14 @@ function generate_password {
 # $4 - reference to the user's answer.
 function ask_question()
 {
-  question_str=$1
-  prev_val=$2
-  val_name=$3
-  answer=''
+    question_str=$1
```
Comment (coverbeck) [inline install_bootstrap:61]: Did you change tabs to spaces or vice-versa, or otherwise do some reformatting?

With exceptions, it's generally best not to add a feature and then also reformat a bunch of the file; it makes it harder to tell what changed.

This comment applies to the whole file, not just this line.

OTOH, if part of your change was just to reformat the indentation, then maybe it's OK.

File: install_bootstrap [hunk 15]
```diff
@@ -181,9 +181,9 @@ done
 
 if [[ "${run_common^^}" = 'Y' ]]; then
     if [[ $(sudo docker ps | grep 'core-' | wc -l) = "0" ]]; then
-      echo "Do you want to run the public-facing gateway nginx server in dev mode or prod? [dev/prod]"
-      common_mode=-1
-      while [[ "${common_mode}" = "-1" ]]; do
+        echo "Do you want to run the public-facing gateway nginx server in dev mode or prod? [dev/prod]"
```
Comment (coverbeck) [inline install_bootstrap]: I'm wondering if we should say recommend prod mode here? What do you think?

**Review by coverbeck — CHANGES_REQUESTED** [review]
I like your changes to the README, I think it's looking a lot better.

I know some of my feedback goes beyond what you changed, but since you're in there, I think it would be good to address.

File: boardwalk/conf/boardwalk.config.template (hunk 4 above)
Comment (jessebrennan) [reply boardwalk/conf/boardwalk.config.template]: Whoops! yes.

File: README.md (hunk 5 above)
Comment (jessebrennan) [reply README.md]: Done.

File: README.md (hunk 7 above)
Comment (jessebrennan) [reply README.md]: No problem. I get your point. I guess the URL just provides a *consistent* way to access the VM.

File: README.md (hunk 11 above)
Comment (jessebrennan) [reply README.md:263]: Sure, moving makes sense. I wasn't sure what to do because the install script doesn't use the word Common, but instead gateway nginx server. I've since added parentheticals to clarify the redundant names.

File: install_bootstrap (hunk 14 above)
Comment (jessebrennan) [reply install_bootstrap:61]: I did all of the reformatting in a separate commit https://github.com/DataBiosphere/cgp-deployment/pull/56/commits/20043e98f3e7306450fd98558d2a051671a8d7e5, that way you would be able to tell the changes apart. The indentation was very inconsistent and it was almost necessary to adjust before adding my own code.

File: README.md (hunk 12 above)
Comment (jessebrennan) [reply README.md]: Sure, I added something.

File: install_bootstrap (hunk 15 above)
Comment (jessebrennan) [reply install_bootstrap]: I added a note that explains the limitations for using dev mode here.

File: README.md [hunk 16]
```diff
@@ -25,7 +25,8 @@ These components are setup with the install process available in this
 repository:
 
 * [Boardwalk](boardwalk/README.md): our file browsing portal on top of
-  Redwood
+  the [Azul
+  Indexer](https://github.com/DataBiosphere/azul/tree/feature/commons).
```
Comment (coverbeck) [inline README.md]: 👍

File: README.md [hunk 17]
```diff
@@ -80,11 +81,12 @@ instance with 60 GB disk has been sufficient.
 
 #### Make a routing to the VM
 
-You will need to make a URL to access the VM. On the AWS console, go to
-route 53 service, click **Hosted zones**. If you already have Hosted
-zone you can use, select it from the list. If not, create one to suit
-your needs. Once the zone is selected, click **Create Record Set** at
-the top. Choose a name you like and use the elastic ip as the value.
+You will need to make a URL for consistent access to the VM. On the AWS
```
Comment (coverbeck) [inline README.md]: I'm not sure I like `consistent`. And in hindsight, I don't think I like `make a URL`. How about something like:

You will need to create a DNS record that maps to your VM.

File: README.md (hunk 10 above)
Comment (coverbeck) [reply README.md]: I think you missed this comment.

File: install_bootstrap [hunk 18]
```diff
@@ -181,7 +181,8 @@ done
 
 if [[ "${run_common^^}" = 'Y' ]]; then
     if [[ $(sudo docker ps | grep 'core-' | wc -l) = "0" ]]; then
-        echo "Do you want to run the public-facing gateway nginx server in dev mode or prod? [dev/prod]"
+        echo "Do you want to run the public-facing gateway nginx server (Common) in dev mode or prod? [dev/prod]"
+        echo "NOTE: dev mode will issue fake certificates, which may be insufficient for developing authentication."
```
Comment (coverbeck) [inline install_bootstrap]: 👍

File: README.md [hunk 19]
```diff
@@ -44,94 +100,213 @@ Open inbound ports on your security group. Use the table below as a guide. Make
...
+`boardwalk_dcc-dashboard-service`, and `boardwalk_boardwalk` from the
+images (see
+[here](https://github.com/DataBiosphere/cgp-deployment/blob/feature/update-readme/boardwalk/README.md#development-mode)
+for more details). In addition, the `nginx` image is built from the
+*nginx-dev* directory. If your work requires real SSL certificates
+during development, it is recommended to set up *Common* in `prod` mode,
+and *Boardwalk* in `dev` mode.
+
+Once the installer completes, the system should be up and running.
+Congratulations! Execute `docker ps` to get an idea of which containers
```
Comment (coverbeck) [inline README.md]: I think you need to run `sudo docker ps`.

**Review by coverbeck — CHANGES_REQUESTED** [review]
Sorry, a couple of more tweaks, including one I missed from before.

File: README.md (hunk 17 above)
Comment (jessebrennan) [reply README.md]: This is better.

File: README.md (hunk 10 above)
Comment (jessebrennan) [reply README.md]: I'll link to https://github.com/DataBiosphere/cgp-data-store as we discussed for more context.

File: README.md [hunk 20]
```diff
@@ -44,94 +100,213 @@ Open inbound ports on your security group. Use the table below as a guide. Make
...
+1. Next you are asked to decide if you want an authorization whitelist.
+
+   If you decide to use a whitelist you will need set it up with
+   [bouncer](https://github.com/ucsc-cgp/bouncer#how-to-use). You will
+   also be prompted for a project name and a contact email which will be
+   used in error messages for users who aren't authorized, but attempt
+   to access the data.
+
+1. You will now have to enter the hostname of dcc-dashboard. This is the
+   domain you made in Rounte 52, above. You need to know the name of the
```
Comment (coverbeck) [inline README.md]: `Route 53`

File: README.md [hunk 21]
```diff
@@ -2,40 +2,96 @@
...
+6. Back in _EC2 Dashboard_ go to *Elastic IPs*. The IP address just
+   created should be in that list. Check it and under *Actions*,
+   *Associate*, choose resource type "Instance", and choose your EC2
+   (e.g., searching by its name).
+7.  <a name="sshconnect"></a>In _EC2 Dashboard_ make your VM active by
+    clicking it. Then click _Connect_ on top. The example in that window
+    shows you how to ssh into your VM from a terminal.
+
+
+#### Make a routing to the VM
```
Comment (coverbeck) [inline README.md]: Can we call this `Create a DNS entry for the VM`, or `Create a DNS entry for the VM in Route 53`?

File: README.md (hunk 20 above)
Comment (coverbeck) [reply README.md]: And can you make `above` be a link to the section? I was looking for it, but had a hard time finding it because the header says `Make a routing to the VM`. I entered a separate suggestion to rename the title, but even so, it would be nice to have a link as well.

**Review by coverbeck — CHANGES_REQUESTED** [review]

File: README.md (hunk 20 above)
Comment (jessebrennan) [reply README.md]: Sure that would help, I agree.

File: README.md (hunk 21 above)
Comment (jessebrennan) [reply README.md]: sounds good to me

**Review by coverbeck — APPROVED** [review]
